infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/multiuser/oauth/KubernetesOidcProviderConfigFactory.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2012-2021 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.multiuser.oauth;
+
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import java.util.Optional;
+import javax.inject.Singleton;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientConfigFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class creates {@link Config} from Kubernetes OIDC token. It does not guarantee that token is
+ * valid.
+ */
+@Singleton
+public class KubernetesOidcProviderConfigFactory extends KubernetesClientConfigFactory {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(KubernetesOidcProviderConfigFactory.class);
+
+  @Override
+  public boolean isPersonalized() {
+    return getToken().isPresent();
+  }
+
+  /**
+   * Builds the OpenShift {@link Config} object based on a default {@link Config} object and token
+   * stored in {@link EnvironmentContext}. It ignores 'workspaceId'.
+   */
+  public Config buildConfig(Config defaultConfig, @Nullable String workspaceId) {
+    return getToken()
+        .map((token) -> new ConfigBuilder(defaultConfig).withOauthToken(token).build())
+        .orElseGet(
+            () -> {
+              LOG.debug("NO TOKEN FOUND. Getting default client config.");
+              return defaultConfig;
+            });
+  }
+
+  private Optional<String> getToken() {
+    return Optional.ofNullable(EnvironmentContext.getCurrent().getSubject().getToken());
```

Comment:
In case we don't have current user, it should return null and we should get default config in `buildConfig`. TBH I'm not sure if it is ok in killing idling workspaces or not. Do you know how it works now? Do we somehow impersonate the user even the user is not doing the request and is not active? Or do we use che-server service account ?

Reply:
To reformulate, do you have concrete idea what would be appropriate to do here ? :)

Reply:
do we even have idling workspace stop for devworkspaces?

Reply:
I don't think we're trying to fake anything - the code just runs on a schedule and doesn't muck around with the environment context in any way.

https://github.com/eclipse-che/che-server/blob/6670c6986b1cb3da5e733e1a0b48db78b393ee3d/wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java#L124

And you're right - `getSubject()` never returns `null` - it returns an anonymous user in the cases like this. I actually have no idea how this works on OpenShift - what identity we impersonate when trying to stop a workspace from outside of any user-initiated interaction. @skabashnyuk do you know how this works?

Reply:
It works because of https://github.com/eclipse-che/che-server/blob/main/infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/OpenShiftStopWorkspaceRoleProvisioner.java

Reply:
in devworkspaces, che-server does not control workspace lifecycle. imho it is responsibility of devworkspace operator to do this job.

Reply:
Can we then at least call the method a different name? Like `getOpenShiftOAuthTokenUsableWithDevworkspacesOnly` or something like that makes it clear that this only works under very specific circumstances?

Reply:
We can, but I'm not sure if it makes sense. First it's not only openshift. Then it's in class `KubernetesOidcProviderConfigFactory`, so it's clear for what setup is should be used. Also there is overridden `isPersonalized` method, that is checking if token is available and javadoc of that method is imho explaining exactly that https://github.com/sparkoo/che-server/blob/gh19703-auth/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientConfigFactory.java#L39

Reply:
still this is private method, we can name it whatever we want and I can put some comments there. However, the 3 things I mentioned in previous comment are imho explaining it quite well. If you have an idea how to make it more transparent,  ideas welcome :)

cc: @metlos

Reply:
@metlos I've merged the PR, it is taking too long. We can fix your concerns in new PR.